io/directory.go

```diff
@@ -573,42 +560,21 @@ func (d *UpgradeableDirectory) getDagService() ipld.DAGService {
 //  sure we make good on the value). Finding the right margin can be tricky
 //  and very dependent on the use case so it might not be worth it.
 func (d *UpgradeableDirectory) RemoveChild(ctx context.Context, name string) error {
-	if err := d.Directory.RemoveChild(ctx, name); err != nil {
-		return err
-	}
-
 	hamtDir, ok := d.Directory.(*HAMTDirectory)
-	if !ok { // BasicDirectory
-		return nil
-	}
-
-	if HAMTShardingSize == 0 || // Option disabled.
-		hamtDir.sizeChange >= 0 { // We haven't reduced the HAMT net size.
-		return nil
-	}
-
-	// We have reduced the directory size, check if it didn't go under
-	// the HAMTShardingSize threshold.
-	belowThreshold, timeoutExceeded := hamtDir.sizeBelowThreshold(EvaluateHAMTTransitionTimeout)
-
-	if timeoutExceeded {
-		// We run out of time before confirming if we're indeed below the
-		// threshold. When in doubt error to not return inconsistent structures.
-		// FIXME: We could allow this to return without error and enforce this
-		//  timeout on a GetNode() call when we need to actually commit to a
-		//  structure/CID. (The downside is that GetNode() doesn't have a
-		//  context argument and we would have to break the API.)
-		return fmt.Errorf("not enought time to fetch shards")
-		// FIXME: Abstract in new error for testing.
-	}
-
-	if belowThreshold { // Switch.
-		basicDir, err := hamtDir.switchToBasic(ctx)
+	if ok {
+		switchToBasic, err := hamtDir.needsToSwitchToBasicDir(ctx, name)
 		if err != nil {
 			return err
 		}
-		d.Directory = basicDir
+
+		if switchToBasic {
+			basicDir, err := hamtDir.switchToBasic(ctx)
+			if err != nil {
+				return err
+			}
+			d.Directory = basicDir
+		}
 	}
 
-	return nil
+	return d.Directory.RemoveChild(ctx, name)
```

Comment:
As https://github.com/ipfs/go-unixfs/pull/103/files#r695554547 we should only set `d.Directory` after we have a fully setup directory. If we do it in the middle like this and then return an error from `d.Directory.RemoveChild` then we have a "corrupted" directory.

Note: when I say corrupted I mean in the sense of the sizes not being reproducible not whether the data is readable, but we should still keep things consistent and easy for the consumer of this library to reason about when possible.